### PR TITLE
http/inbound: apply TTL to context

### PR DESCRIPTION
**Pull request by kriskowal**
This applies the TTL from the inbound request to the inbound request context.

### CODE REVIEW

**Comment by kriskowal** [comment]
r @yarpc/golang

File: transport/transporttest/context.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+package transporttest
+
+import (
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+// ContextMatcher TODO
```
Comment (abhinav) [inline transport/transporttest/context.go]: We can fill the docs now

File: transport/transporttest/context.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	if ttl > maxTTL || ttl < minTTL {
+		c.t.Logf("TTL out of expected bounds: %v < %v < %v", minTTL, ttl, maxTTL)
+		return false
+	}
+
+	return true
+}
+
+func (c *ContextMatcher) String() string {
+	return "lol"
```
Comment (prashantv) [inline transport/transporttest/context.go]: lol. do the matchers need a `String` method? If so might want to do something like
```go
return fmt.Sprintf("ContextMatcher(%v ± %v", c.ttl, c.ttlDelta)
```

**Comment by prashantv** [comment]
Don't see any changes to actually set the TTLs -- is that already being done? Does the `TTL` field in the `transport.Request` already be set?

**Comment by kriskowal** [comment]
@prashantv Yes, the TTL is already being read from the request headers and placed on the transport request. Only missing piece was communicating it to the context.

File: transport/transporttest/context.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+
+// Matches a context against an expected context, returning true only if the
+// given object is a context with a deadline that is now, plus the expected
+// TTL, plus or minus some tolerance.
+func (c *ContextMatcher) Matches(got interface{}) bool {
+	ctx, ok := got.(context.Context)
+	if !ok {
+		fmt.Println("no context")
```
Comment (prashantv) [inline transport/transporttest/context.go]: you use `c.t.Logf` but `Println` here -- intentional?

File: transport/transporttest/context.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		return false
+	}
+
+	d, ok := ctx.Deadline()
+	if !ok {
+		fmt.Println("no context deadline")
```
Comment (prashantv) [inline transport/transporttest/context.go]: same as above

**Comment by prashantv** [comment]
lgtm with comment on prints

File: transport/transporttest/context.go (hunk 4 above)
Comment (kriskowal) [inline transport/transporttest/context.go]: We were running into this during debugging and it wasn’t reported to the console. I would be inclined to remove them save that they were useful.

Comment (prashantv) [inline transport/transporttest/context.go]: yep they can be pretty useful, but the `Logf` should be reported on test failure so that should be enough? Not sure I understand why it's `Println` here but `t.Logf` below

File: transport/transporttest/context.go (hunk 3 above)
Comment (abhinav) [inline transport/transporttest/context.go]: +1. Delete the printlns. Add log statements on the test if necessary.

File: transport/transporttest/context.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	if ttl > maxTTL || ttl < minTTL {
+		c.t.Logf("TTL out of expected bounds: %v < %v < %v", minTTL, ttl, maxTTL)
+		return false
+	}
+
+	return true
+}
+
+func (c *ContextMatcher) String() string {
+	return fmt.Sprintf("ContextMatcher(TTL:%v±%v)", c.ttl, c.ttlDelta)
```
Comment (abhinav) [inline transport/transporttest/context.go]: heh unicode

File: transport/transporttest/context.go (hunk 3 above)
Comment (breerly) [inline transport/transporttest/context.go]: I would switch these to a log statement, then run the tests with `go -v` to see that it shows up

File: transport/transporttest/context.go (hunk 4 above)
Comment (breerly) [inline transport/transporttest/context.go]: If you're trying to create verbose output for debugging on non-failure cases, try using log and running tests with `go test -v`...

File: transport/http/handler_test.go [hunk 6]
```diff
@@ -135,7 +135,7 @@ func TestHandlerInternalFailure(t *testing.T) {
 
 	rpcHandler := transporttest.NewMockHandler(mockCtrl)
 	rpcHandler.EXPECT().Handle(
-		gomock.Any(),
+		transporttest.NewContextMatcher(t, time.Second),
```
Comment (breerly) [inline transport/http/handler_test.go]: What are you testing here? Just passing in a `time.Second`? I would expect some interaction with TTL here

**Comment by breerly** [comment]
I think we need some more tests to nail down this behavior.

**Comment by kriskowal** [comment]
@breerly added a happy path test. This is a unit test that verifies that the handler propagates the request TTL header into the context and request object, among other properties.

File: transport/http/handler_test.go (hunk 6 above)
Comment (abhinav) [inline transport/http/handler_test.go]: The context matcher will test that the Context this function is called with has a TTL of one second give or take a few milliseconds.

File: transport/http/handler_test.go [hunk 7]
```diff
@@ -38,6 +38,47 @@ import (
...
+
+	req := &http.Request{
+		Method: "POST",
+		Header: headers,
+		Body:   ioutil.NopCloser(bytes.NewReader([]byte("Nyuck Nyuck"))),
+	}
+	rw := httptest.NewRecorder()
+	httpHandler.ServeHTTP(rw, req)
+	code := rw.Code
+	assert.True(t, code == 200, "expected 200 code")
```
Comment (abhinav) [inline transport/http/handler_test.go]: Use `assert.Equal`

File: transport/http/handler_test.go [hunk 8]
```diff
@@ -38,6 +38,47 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestHandlerSucces(t *testing.T) {
```
Comment (breerly) [inline transport/http/handler_test.go]: What about TChannel?

Comment (abhinav) [inline transport/http/handler_test.go]: That'll be in a separate PR. This PR is for HTTP only.

Comment (breerly) [inline transport/http/handler_test.go]: Aha - I should get better at reading PR titles 👍

**Comment by breerly** [comment]
:shipit:

**Comment by breerly** [comment]
Instead of writing a test for HTTP and TChannel separately, consider writing a single integration test with Crossdock.
